File: gclib/src/main/java/com/github/guqt178/helper/ActivityResultHelper.java
package com.github.guqt178.helper;

import android.content.Context;
import android.content.Intent;
import android.support.v4.app.FragmentActivity;
import android.support.v4.app.FragmentManager;

/**
 * 借助fragment优雅地申请权限和处理 onActivityResult
 * @see RouterFragment
 * @see "https://blog.csdn.net/gdutxiaoxu/article/details/86498647"
 */
public class ActivityResultHelper {

    private static final String TAG = "ActivityLauncher";
    private Context mContext;
    private RouterFragment mRouterFragment;

    public static ActivityResultHelper init(FragmentActivity activity) {
        return new ActivityResultHelper(activity);
    }

    private ActivityResultHelper(FragmentActivity activity) {
        mContext = activity;
        mRouterFragment = getRouterFragment(activity);
    }

    private RouterFragment getRouterFragment(FragmentActivity activity) {
        RouterFragment routerFragment = findRouterFragment(activity);
        if (routerFragment == null) {
            routerFragment = RouterFragment.newInstance();
            FragmentManager fragmentManager = activity.getSupportFragmentManager();
            fragmentManager
                    .beginTransaction()
                    .add(routerFragment, TAG)
                    .commitAllowingStateLoss();
            fragmentManager.executePendingTransactions();
        }
        return routerFragment;
    }

    private RouterFragment findRouterFragment(FragmentActivity activity) {
        return (RouterFragment) activity.getSupportFragmentManager().findFragmentByTag(TAG);
    }

    public void startActivityForResult(Class<?> clazz, OnResult callback) {
        Intent intent = new Intent(mContext, clazz);
        startActivityForResult(intent, callback);
    }

    public void startActivityForResult(Intent intent, OnResult callback) {
        mRouterFragment.startActivityForResult(intent, callback);
    }

}

File: libs/gclib/src/main/kotlin/com/github/guqt178/utils/result/ActivityResultHelper.kt
package com.github.guqt178.utils.result

import android.content.Context
import android.content.Intent
import android.support.v4.app.FragmentActivity

/**
 * 借助fragment优雅地申请权限和处理 onActivityResult
 * @see RouterFragment
 *
 * @see "https://blog.csdn.net/gdutxiaoxu/article/details/86498647"
 */
class ActivityResultHelper private constructor(activity: FragmentActivity) {
  private val mContext: Context
  private val mRouterFragment: RouterFragment?
  private fun getRouterFragment(activity: FragmentActivity): RouterFragment? {
    var routerFragment: RouterFragment? = findRouterFragment(activity)
    if (routerFragment == null) {
      routerFragment = RouterFragment.newInstance()
      val fragmentManager = activity.supportFragmentManager
      fragmentManager
          .beginTransaction()
          .add(
              routerFragment,
              TAG
          )
          .commitAllowingStateLoss()
      fragmentManager.executePendingTransactions()
    }
    return routerFragment
  }

  private fun findRouterFragment(activity: FragmentActivity): RouterFragment {
    return activity.supportFragmentManager
        .findFragmentByTag(
            TAG
        ) as RouterFragment
  }

  fun startActivityForResult(
    clazz: Class<*>?,
    callback: OnResult?
  ) {
    val intent = Intent(mContext, clazz)
    startActivityForResult(intent, callback)
  }

  fun startActivityForResult(
    intent: Intent?,
    callback: OnResult?
  ) {
    mRouterFragment!!.startActivityForResult(intent, callback)
  }

  companion object {
    private const val TAG = "ActivityLauncher"
    fun init(activity: FragmentActivity): ActivityResultHelper {
      return ActivityResultHelper(
          activity
      )
    }
  }

  init {
    mContext = activity
    mRouterFragment = getRouterFragment(activity)
  }
}